From fdb0a0bc03de01c18d9f209c6cf0f5ae2aec4840 Mon Sep 17 00:00:00 2001
Message-Id: <fdb0a0bc03de01c18d9f209c6cf0f5ae2aec4840.1298580377.git.jdenemar@redhat.com>
From: Roopa Prabhu <roprabhu@cisco.com>
Date: Tue, 22 Feb 2011 09:46:18 -0700
Subject: [PATCH] 802.1Qbh: Delay IFF_UP'ing interface until migration final stage

https://bugzilla.redhat.com/show_bug.cgi?id=678826

Current code does an IFF_UP on a 8021Qbh interface immediately after a port
profile set. This is ok in most cases except when its the migration prepare
stage. During migration we want to postpone IFF_UP'ing the interface on the
destination host until the source host has disassociated the interface.
This patch moves IFF_UP of the interface to the final stage of migration.
The motivation for this change is to postpone any addr registrations on the
destination host until the source host has done the addr deregistrations.

While at it, for symmetry with associate move ifDown of a 8021Qbh interface
to before disassociate
(cherry picked from commit dfd39ccda8d0347636d92b89bf569b1dc9924ea6)
---
 src/util/macvtap.c |   13 ++++++-------
 1 files changed, 6 insertions(+), 7 deletions(-)

diff --git a/src/util/macvtap.c b/src/util/macvtap.c
index 96df301..76910e1 100644
--- a/src/util/macvtap.c
+++ b/src/util/macvtap.c
@@ -1470,8 +1470,6 @@ doPortProfileOp8021Qbh(const char *ifname,
                                   NULL,
                                   vf,
                                   PORT_REQUEST_DISASSOCIATE);
-        if (!rc)
-            ifaceUp(ifname);
         break;
 
     case DISASSOCIATE:
@@ -1484,7 +1482,6 @@ doPortProfileOp8021Qbh(const char *ifname,
                                    NULL,
                                    vf,
                                    PORT_REQUEST_DISASSOCIATE);
-        ifaceDown(ifname);
         break;
 
     default:
@@ -1550,10 +1547,11 @@ vpAssociatePortProfileId(const char *macvtap_ifname,
 
     case VIR_VIRTUALPORT_8021QBH:
         /* avoid associating twice */
-        if (vmOp == VIR_VM_OP_MIGRATE_IN_FINISH)
-            break;
-        rc = doPortProfileOp8021Qbh(linkdev, macvtap_macaddr,
-                                    virtPort, vmuuid, ASSOCIATE);
+        if (vmOp != VIR_VM_OP_MIGRATE_IN_FINISH)
+            rc = doPortProfileOp8021Qbh(linkdev, macvtap_macaddr,
+                                        virtPort, vmuuid, ASSOCIATE);
+        if (vmOp != VIR_VM_OP_MIGRATE_IN_START && !rc)
+            ifaceUp(linkdev);
         break;
     }
 
@@ -1600,6 +1598,7 @@ vpDisassociatePortProfileId(const char *macvtap_ifname,
         /* avoid disassociating twice */
         if (vmOp == VIR_VM_OP_MIGRATE_IN_FINISH)
             break;
+        ifaceDown(linkdev);
         rc = doPortProfileOp8021Qbh(linkdev, macvtap_macaddr,
                                     virtPort, NULL, DISASSOCIATE);
         break;
-- 
1.7.4.1

